lib/broadway/rate_limiter.ex

```diff
@@ -0,0 +1,77 @@
+defmodule Broadway.RateLimiter do
+  @moduledoc false
+
+  use GenServer
+  @row_name :rate_limit_counter
+
+  def start_link(opts) do
+    case Keyword.fetch!(opts, :rate_limiting) do
+      # If we don't have rate limiting options, we don't even need to start this rate
+      # limiter process.
+      nil ->
+        :ignore
+
+      rate_limiting_opts ->
+        name = Keyword.fetch!(opts, :name)
+        GenServer.start_link(__MODULE__, {name, rate_limiting_opts})
+    end
+  end
+
+  def decrease_counter_or_rate_limit(broadway_name) do
+    # We're decrementing by 1. We're setting a threshold of 0, which means that
+    # if the counter would go below 0, it's reset to set_value (-1) instead.
+    # This ensures that the first time we go below 0, we reset to -1 and return -1.
+    # Then every other time we're going to decrement from -1 to -2, reset to -1
+    # because it's below the threshold, and return -1 again. So -1 means that we
+    # are below the rate limiting.
+
+    update_op = {_position = 2, _incr = -1, _threshold = 0, _set_value = -1}
+
+    case :ets.update_counter(table_name(broadway_name), @row_name, update_op) do
+      -1 -> :rate_limited
+      _other -> :ok
+    end
+  end
+
+  def get_currently_allowed(broadway_name) do
+    :ets.lookup_element(table_name(broadway_name), @row_name, 2)
+  end
+
+  @impl true
+  def init({name, rate_limiting_opts}) do
+    interval = Keyword.fetch!(rate_limiting_opts, :interval)
+    allowed = Keyword.fetch!(rate_limiting_opts, :allowed_messages)
+
+    table_name = table_name(name)
+    _ets = :ets.new(table_name, [:named_table, :public, :set])
+    :ets.insert(table_name, {@row_name, allowed})
+
+    :timer.send_interval(interval, {:reset_limit, allowed})
```

Comment:
Use Process.send_after/3 instead. :timer is a singleton process that can become a bottleneck if overused.

Reply:
Done.